File: leetcode/all-divisions-with-the-highest-score-of-a-binary-array.cpp
// https://leetcode.com/problems/all-divisions-with-the-highest-score-of-a-binary-array/
#include "bits/stdc++.h"
using namespace std;

class Solution {
public:
  vector<int> maxScoreIndices(vector<int> &nums) {
    int n = nums.size();
    int zeros = 0;
    int ones = 0;
    for (auto &x : nums) {
      if (x == 1)
        ones += x;
    }
    int score = ones;
    vector<int> res;

    res.push_back(0);

    for (int i = 0; i < n; i++) {
      int x = nums[i];

      switch (x) {
      case 0:
        zeros++;
        break;
      default:
        ones--;
        break;
      }

      int total = zeros + ones;

      if (total == score) {
        res.push_back(i + 1);
      } else if (total > score) {
        score = total;
        res.clear();
        res.push_back(i + 1);
      }
    }

    return res;
  }
};

File: leetcode/scala/src/main/scala/all-divisions-with-the-highest-score-of-a-binary-array.scala
// https://leetcode.com/problems/all-divisions-with-the-highest-score-of-a-binary-array/
package maxScoreIndices
import scala.collection.mutable

object Solution {
  def maxScoreIndices(nums: Array[Int]): List[Int] = {
    var zeros = 0
    var ones = nums.sum
    var score = ones
    val res = mutable.ListBuffer[Int]()
    res.append(0)

    (0 to nums.size - 1).foreach { i =>
      val x = nums(i)

      x match {
        case 0 =>
          zeros += 1
        case _ =>
          ones -= 1
      }

      zeros + ones match {
        case v if v == score =>
          res += i
        case v if v > score =>
          score = v
          res.clear()
          res += i
        case _ =>
      }
    }
    res.toList
  }
}
